### PR TITLE
[tensor] design DistSpec and DistSpecManager for ColoTensor

### PR DESCRIPTION
**Pull request by ver217**
1. design DistSpec and DistSpecManager for ColoTensor
2. update current ops, besides loss
3. skip `test_model.py`

### CODE REVIEW

File: colossalai/tensor/colo_tensor.py [hunk 1]
```diff
@@ -127,85 +132,21 @@ def torch_tensor(self) -> torch.Tensor:
...
-        dim = self._get_gather_dim()
-        out_tensor = torch.cat(tensor_list, dim=dim).contiguous()
-
-        return out_tensor
-
-    def is_gathered(self) -> bool:
-        return self._shard_pattern == ShardPattern.NA
+    def set_spec(self, spec: TensorSpec, shard: bool = True, init_pg: bool = False) -> None:
+        spec = copy(spec)
+        if init_pg:
```
Comment (Wesley-Jzy) [inline colossalai/tensor/colo_tensor.py:137]: What is init_pg?

File: colossalai/tensor/colo_tensor.py [hunk 2]
```diff
@@ -296,3 +228,9 @@ def __mul__(self, other) -> "ColoTensor":
             raise TypeError(f'{type(other)} is not supported in ColoTensor __mul__')
 
     __rmul__ = __mul__
+
+    def to_dist_spec(self, dist_spec: _DistSpec) -> None:
+        self._torch_tensor = DistSpecManager.handle_trans_spec(self.torch_tensor(), self.spec.dist_spec, dist_spec)
+        # self._torch_tensor.requires_grad = self._requires_grad
+        self._size = self._torch_tensor.size()
```
Comment (Wesley-Jzy) [inline colossalai/tensor/colo_tensor.py:235]: I think the attributes in current ColoTensor need to be defined clearly, like size,device.... Maybe they mean either local or global tensor info now.

File: colossalai/tensor/colo_tensor.py (hunk 1 above)
Comment (ver217) [reply colossalai/tensor/colo_tensor.py:137]: Some tensor is initialized with wrong process group. We cannot handle process group change now, so we just copy new process group to the previous DistSpec. It's a temporary design and I may change this function in the future.

Comment (ver217) [reply colossalai/tensor/colo_tensor.py]: I already removed this arg.

File: colossalai/tensor/colo_tensor.py (hunk 2 above)
Comment (ver217) [reply colossalai/tensor/colo_tensor.py:232]: Shall we add two methods like `is_global()` and `to_global()` ? In this way, we assume ColoTensor is local by default. If you have any good ideas, please let me know.

File: colossalai/tensor/colo_tensor.py (hunk 1 above)
Comment (Wesley-Jzy) [reply colossalai/tensor/colo_tensor.py]: ok

File: colossalai/tensor/colo_tensor.py (hunk 2 above)
Comment (Wesley-Jzy) [reply colossalai/tensor/colo_tensor.py:232]: Make sense. I added a global_torch_tensor() in ColoTensor before to get tensors in global view, but it only serves for activations. Maybe we can re-implement it with dictspec.

Comment (ver217) [reply colossalai/tensor/colo_tensor.py:232]: We can do this in the next PR.

**Comment by ver217** [comment]
All tests passed.
image

File: colossalai/tensor/__init__.py [hunk 3]
```diff
@@ -1,13 +1,14 @@
-from .spec import ComputePattern, ParallelAction, TensorSpec, ShardPattern
+from .spec import ComputePattern, ParallelAction, TensorSpec
 from .op_wrapper import (
     colo_op_impl,)
 from .colo_tensor import ColoTensor
 from .colo_parameter import ColoParameter
 from .utils import convert_parameter, named_params_with_colotensor
 from ._ops import *
 from .optim.colo_optimizer import ColoOptimizer
+from . import dist_spec
```
Comment (feifeibear) [inline colossalai/tensor/__init__.py:9]: why not name it as DistSpec?

Comment (feifeibear) [reply colossalai/tensor/__init__.py:9]: Or DistSignature? to be consistent with our design doc?

Comment (ver217) [reply colossalai/tensor/__init__.py:9]: It's a package not class. I just use package to implement the "namespace" in C++.

File: colossalai/tensor/colo_tensor.py [hunk 4]
```diff
@@ -127,85 +132,17 @@ def torch_tensor(self) -> torch.Tensor:
...
-    def is_gathered(self) -> bool:
-        return self._shard_pattern == ShardPattern.NA
+    def set_spec(self, spec: TensorSpec) -> None:
+        spec = copy(spec)
+        self.to_dist_spec(spec.dist_spec)
+        self._spec = spec
 
     def has_spec(self) -> bool:
-        return self._shard_spec is not None and self._shard_spec.num_action > 0
+        return self._spec is not None and self._spec.num_action > 0
```
Comment (feifeibear) [inline colossalai/tensor/colo_tensor.py:141]: we'd better set _spec as a dummy one by default.
I think making it an optional variable may bring some trouble. According to our experiments in ShardTensor design.

File: colossalai/tensor/__init__.py (hunk 3 above)
Comment (feifeibear) [reply colossalai/tensor/__init__.py:9]: OK,that's fine

Comment (ver217) [reply colossalai/tensor/__init__.py:9]: I've updated this part in our design doc. I think current implementation is simpler and it's easy to comprehend for users.

File: colossalai/tensor/_ops/linear.py [hunk 5]
```diff
@@ -4,72 +4,50 @@
...
-    elif input_tensor.shard_pattern == ShardPattern.Col:
-        # Splited by 1Dcol
-        assert input_tensor.shape[-1] == weight.size(-1), \
-        'Invalid shapes in 1Drow forward: input={}, weight={}. Expected last dim of input {}.'.format(
-        input_tensor.shape, weight.size, weight.size(-1))
-        input_per_partition = input_tensor.torch_tensor()
-    else:
-        raise NotImplementedError
+    input_tensor.to_dist_spec(
+        dist_spec.shard(weight.spec.dist_spec.process_group, [-1], [weight.spec.dist_spec.process_group.size()]))
```
Comment (feifeibear) [inline colossalai/tensor/_ops/linear.py:17]: The nesting naming is confusion....
a file.and sepc.py also a file containing dist_spec? But dist_spec.py is also a file.

Comment (feifeibear) [reply colossalai/tensor/_ops/linear.py:17]: why not just make a file named api.py?

File: colossalai/tensor/colo_tensor.py (hunk 4 above)
Comment (ver217) [reply colossalai/tensor/colo_tensor.py:141]: Oh, yes. Actually our default value of _spec is not None. This code snippet is outdated and I forgot to update.

File: colossalai/tensor/_ops/linear.py (hunk 5 above)
Comment (feifeibear) [reply colossalai/tensor/_ops/linear.py:17]: In short, I think the following usage looks better to developers.
weight.spec.get_process_group()

File: colossalai/tensor/colo_tensor.py (hunk 4 above)
Comment (ver217) [reply colossalai/tensor/colo_tensor.py]: updated

File: colossalai/tensor/_ops/linear.py (hunk 5 above)
Comment (Wesley-Jzy) [reply colossalai/tensor/_ops/linear.py:17]: I think the first spec is TensorSpec and the second one is DistSpec. Do we really neead TensorSpec abs?

Comment (ver217) [reply colossalai/tensor/_ops/linear.py:17]: Flattern the folder structure now.

Comment (ver217) [reply colossalai/tensor/_ops/linear.py:17]: TensorSpec includes a action list now.

Comment (ver217) [reply colossalai/tensor/_ops/linear.py:17]: If we move the action list to op rather than parameter, then TensorSpec and DistSpec can be merged into one.

Comment (Wesley-Jzy) [reply colossalai/tensor/_ops/linear.py:17]: It's actually introduced by ComputePattern bound to weight.

Comment (Wesley-Jzy) [reply colossalai/tensor/_ops/linear.py:17]: Use the weight.spec.get_process_group() mentioned by @feifeibear or add a TODO for ComputePattern code refining.

**Review by Wesley-Jzy — APPROVED** [review]

File: colossalai/tensor/_ops/linear.py (hunk 5 above)
Comment (ver217) [reply colossalai/tensor/_ops/linear.py]: `spec.get_process_group()` added.
